// typegraphql-prisma/generator/resolvers/helpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const helpers_1 = require("../helpers");
function generateCrudResolverClassMethodDeclaration(operationKind, actionName, method, argsTypeName, collectionName, modelNames, mapping, options) {
    const returnTSType = helpers_1.getFieldTSType(method.outputType, modelNames);
    return {
        name: options.useOriginalMapping
            ? method.name
            : helpers_1.getMappedActionName(actionName, method.name, mapping),
        isAsync: true,
        returnType: `Promise<${returnTSType}>`,
        decorators: [
            {
                name: `TypeGraphQL.${operationKind}`,
                arguments: [
                    `_returns => ${helpers_1.getTypeGraphQLType(method.outputType, modelNames)}`,
                    `{
            nullable: ${!method.outputType.isRequired},
            description: undefined
          }`,
                ],
            },
        ],
        parameters: [
            ...(actionName === "aggregate"
                ? []
                : [
                    {
                        name: "ctx",
                        // TODO: import custom `ContextType`
                        type: "any",
                        decorators: [{ name: "TypeGraphQL.Ctx", arguments: [] }],
                    },
                ]),
            ...(!argsTypeName
                ? []
                : [
                    {
                        name: "args",
                        type: argsTypeName,
                        decorators: [{ name: "TypeGraphQL.Args", arguments: [] }],
                    },
                ]),
        ],
        statements: actionName === "aggregate"
            ? [
                // it will expose field resolvers automatically
                `return new ${returnTSType}();`,
            ]
            : [
                `return ctx.prisma.${collectionName}.${actionName}(${argsTypeName ? "args" : ""});`,
            ],
    };
}
exports.generateCrudResolverClassMethodDeclaration = generateCrudResolverClassMethodDeclaration;
